OpenRA.Mods.Common/Traits/Player/ProductionQueue.cs

```diff
@@ -293,12 +310,28 @@ public virtual IEnumerable<ActorInfo> BuildableItems()
 				return Enumerable.Empty<ActorInfo>();
 			if (!Enabled)
 				return Enumerable.Empty<ActorInfo>();
-			if (developerMode.AllTech)
+			if (!Info.PayUpFront && developerMode.AllTech)
 				return Producible.Keys;
-
+			if (Info.PayUpFront && developerMode.AllTech)
+				return Producible.Keys.Where(a => GetProductionCost(a) <= playerResources.GetCashAndResources() || IsInQueue(a));
+			if (Info.PayUpFront)
+				return buildableProducibles.Where(a => GetProductionCost(a) <= playerResources.GetCashAndResources() || IsInQueue(a));
 			return buildableProducibles;
 		}
 
+		public virtual bool AnyItemsToBuild()
+		{
+			if (productionTraits.Length > 0 && productionTraits.All(p => p.IsTraitDisabled))
+				return false;
+			if (!Enabled)
+				return false;
+			if (developerMode.AllTech && Producible.Keys.Count != 0)
+				return true;
+			if (buildableProducibles.Any())
+				return true;
+			return false;
```

Comment:
```suggestion
			return Enabled
				&& (productionTraits.Length <= 0 || productionTraits.Any(p => p.IsTraitDisabled))
				&& ((developerMode.AllTech && Producible.Keys.Count != 0) || buildableProducibles.Any());
```

this can be shortened

Reply:
also as BuildableItems is virtual all of its overrides should be handled. Though looking at them I feel like you can remove the overrides, they aren't adding anything of value. Git blame doesn't reveal anything, though granted I haven't tested the change I'm proposing